File: JavaScript/0212. Word Search II.js
// Given a 2D board and a list of words from the dictionary, find all words in the board.
//
// Each word must be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.
//
// Example:
//
// Input:
//   words = ["oath","pea","eat","rain"] and board =
//   [
//     ['o','a','a','n'],
//     ['e','t','a','e'],
//     ['i','h','k','r'],
//     ['i','f','l','v']
//   ]
//
// Output: ["eat","oath"]
//
// Note:
// You may assume that all inputs are consist of lowercase letters a-z.

/**
 * @param {character[][]} board
 * @param {string[]} words
 * @return {string[]}
 */

/** Backtracking + Trie */
const findWords = (board, words) => {
  const dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]];
  let res = [];

  const buildTrie = () => {
    const root = {};
    for (const w of words) {
      let node = root;
      for (const c of w) {
        if (node[c] == null) node[c] = {};
        node = node[c];
      }
      node.word = w;
    }
    return root;
  };

  const search = (node, x, y) => {
    if (node.word != null) {
      res.push(node.word);
      node.word = null; // make sure only print one time for each word
    }

    if (x < 0 || x >= board.length || y < 0 || y >= board[0].length) return;
    if (node[board[x][y]] == null) return;

    const c = board[x][y];
    board[x][y] = '#'; // Mark visited
    for (const [dx, dy] of dirs) {
      const i = x + dx;
      const j = y + dy;
      search(node[c], i, j);
    }
    board[x][y] = c; // Reset
  };

  const root = buildTrie();
  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[0].length; j++) {
      search(root, i, j);
    }
  }
  return res;
};

File: Python/0212. Word Search II.py
# Given an m x n board of characters and a list of strings words, return all words on the board.
# Each word must be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.
#
# Example 1:
#
# Input: board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]], words = ["oath","pea","eat","rain"]
# Output: ["eat","oath"]
#
# Example 2:
#
#
# Input: board = [["a","b"],["c","d"]], words = ["abcb"]
# Output: []
#
# Constraints:
#
# m == board.length
# n == board[i].length
# 1 <= m, n <= 12
# board[i][j] is a lowercase English letter.
# 1 <= words.length <= 3 * 104
# 1 <= words[i].length <= 10
# words[i] consists of lowercase English letters.
# All the strings of words are unique.


# Backtracking + Trie
class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        m, n = len(board), len(board[0])
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        res = []

        def build_trie():
            root = {}
            for w in words:
                node = root
                for c in w:
                    # Method 1
                    # if c not in node:
                    #     node[c] = {}
                    # node = node[c]

                    # Method 2
                    node = node.setdefault(c, {})
                node["word"] = w
            return root

        def search(root, x, y):
            c = board[x][y]
            node = root[c]

            # Optimization: incrementally remove the matched leaf node in Trie.
            if not node:
                root.pop(c)

            w = node.pop("word", None)
            if w:
                res.append(w)

            board[x][y] = "#"
            for (dx, dy) in dirs:
                i, j = x + dx, y + dy
                if 0 <= i < m and 0 <= j < n and board[i][j] in node:
                    search(node, i, j)
            board[x][y] = c

        root = build_trie()
        for i in range(m):
            for j in range(n):
                if board[i][j] in root:
                    search(root, i, j)
        return res
